File: history_line.go
package mj

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

const timeFormat = "2006-01-02T15:04:05-0700"

// ---------------------------------------------------------------------
// Type definitions
// ---------------------------------------------------------------------

// HistoryLine is a single record in mahjongg history.
// There is a file named "history" in ~/.local/share/gnome-mahjongg
// which contains a line for each completed game. This line contains:
//  1. A date and time stamp in the format yyyy-mm-ddTHH:MM:SS-9999
//  2. The name of the level (e.g., "easy", "overpass", etc.)
//  3. The number of seconds it took to complete the game.
//
// For example:
//
//	2023-01-04T22:12:03-0500 easy 209
//	2023-01-04T22:15:04-0500 easy 169
//	2023-01-04T22:20:38-0500 overpass 269
//	2023-01-06T14:31:35-0500 overpass 453
//	2023-01-06T16:36:48-0500 tictactoe 228
//	2023-01-06T16:42:17-0500 confounding 252
//	2023-01-06T17:01:03-0500 overpass 299
//	2023-01-06T17:07:52-0500 ziggurat 285
//	2023-01-06T17:13:49-0500 confounding 288
type HistoryLine struct {
	GameDateTime time.Time
	LevelName    string
	Seconds      int
}

// ---------------------------------------------------------------------
// Constructors
// ---------------------------------------------------------------------

// NewHistoryLine creates a new HistoryLine structure from a line in
// the mahjongg history file
func NewHistoryLine(line string) (HistoryLine, error) {

	tokens := strings.Split(line, " ")
	historyLine := HistoryLine{}
	gameDateTime, err := time.Parse(timeFormat, tokens[0])
	if err != nil {
		return historyLine, err
	}

	historyLine.GameDateTime = gameDateTime
	historyLine.LevelName = tokens[1]
	historyLine.Seconds, _ = strconv.Atoi(tokens[2])

	return historyLine, err
}

// ---------------------------------------------------------------------
// Methods
// ---------------------------------------------------------------------

// TimeDate returns the number of seconds found in this record in the
// format hh:mm:ss (MM/DD/YYYY)
func (hl HistoryLine) TimeDate() string {
	part1 := FormatTime(hl.Seconds)
	part2 := hl.GameDateTime.Format(time.RFC3339)[:10]
	return fmt.Sprintf("%s (%s)", part1, part2)
}

// String returns a string representation of the type
func (hl HistoryLine) String() string {
	var parts []string
	parts = append(parts, fmt.Sprintf("GameDateTime=%q", hl.GameDateTime.Format(timeFormat)))
	parts = append(parts, fmt.Sprintf("LevelName=%q", hl.LevelName))
	parts = append(parts, fmt.Sprintf("Seconds=%d", hl.Seconds))
	return strings.Join(parts, ", ")
}

// ---------------------------------------------------------------------
// Functions
// ---------------------------------------------------------------------

// DateString returns the time as a date-only string in yyyy-mm-dd format
func DateString(t time.Time) string {
	return fmt.Sprintf("%04d-%02d-%02d", t.Year(), t.Month(), t.Day())
}

// FormatTime creates a string with hh:mm:ss from the specified number of seconds
func FormatTime(seconds int) string {
	mm := int(seconds / 60)
	ss := seconds % 60
	timeString := fmt.Sprintf("%02d:%02d", mm, ss)
	if mm >= 60 {
		hh := int(mm / 60)
		mm = mm % 60
		timeString = fmt.Sprintf("%02d:%02d:%02d", hh, mm, ss)
	}
	return timeString
}

File: python/mj/history_line.py
from datetime import datetime


class HistoryLine:
    """ A single record in mahjongg history
    """

    def __init__(self, line):
        """Creates a new History record from a line from the history file.
        The line has the format (datetime level seconds)."""
        self._line = line
        tokens = line.split()
        self._game_date: datetime = datetime.strptime(tokens[0], '%Y-%m-%dT%H:%M:%S%z')
        self._level_name: str = tokens[1]
        self._seconds: int = int(tokens[2])

    def __eq__(self, other):
        return str(self) == str(other)

    @staticmethod
    def format_time(seconds) -> str:
        """Creates a string with hh:mm:ss from the specified number of seconds"""
        seconds = int(seconds)
        mm = seconds // 60
        ss = seconds % 60
        time_string = f"{mm:02d}:{ss:02d}"
        if mm >= 60:
            hh = mm // 60
            mm = mm % 60
            time_string = f"{hh:02d}:{mm:02d}:{ss:02d}"
        return time_string

    @property
    def game_date(self) -> datetime:
        """Returns the game date"""
        return self._game_date

    @property
    def level_name(self) -> str:
        """Returns the level name"""
        return self._level_name

    @property
    def line(self) -> str:
        return self._line

    def __repr__(self) -> str:
        output = "HistoryLine("
        output += ")"
        output = f"""{__class__.__name__}("{self.line}")"""
        return output

    @property
    def seconds(self) -> int:
        """Returns the number of seconds in this game"""
        return self._seconds

    def seconds_as_mmss(self) -> str:
        """Converts a number of seconds into hh:mm:ss"""
        return HistoryLine.format_time(self.seconds)

    def __str__(self) -> str:
        """Converts this object into a string representation"""
        tokens = (
            self.game_date.date().isoformat(),
            self.level_name,
            self.seconds_as_mmss()
        )
        output = " ".join(tokens)
        return output

    def timedate(self) -> str:
        """Returns the number of seconds found in this record
        in the format hh:mm:ss (MM/DD/YYYY)"""
        return f"{self.seconds_as_mmss()} ({self.game_date.date().isoformat()})"
